### PR TITLE
workflows: add Linux CI using github actions

### PR DESCRIPTION
**Pull request by iMichka**
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

File: .github/workflows/linux-pr-tests.yml
Comment (carlocab) [inline .github/workflows/linux-pr-tests.yml:95]: Not doing this here?

https://github.com/Homebrew/homebrew-core/blob/7ec2d116a180efd718f0db47e087cf0fe5aa2302/.github/workflows/tests.yml#L150-L161

**Comment by iMichka** [comment]
For some context, my bottling plan is the following:
- Use the Linux dispatch bottle workflow to start bottling low-level formulae, one by one
- In parallel continue syncing both core repositories.
- Once a formula is bottled here, any new update should be run on Linux CI in this repo.

The test workflow should only rebuild a Linux bottle in the case there was previously a Linux bottle for that formula. Else it should just ignore that part of the build. Can we achieve this with test-bot?

I excluded a first list of builds I know might take more than 6 hours. We will have the same list in the new self-hosted workflow, once I have a VM setup. Nevertheless, this PR can be merged when ready, as it should just let the workflow run but actually do nothing, as no previous Linux bottle exist. (if we are able to handle this case in test-bot).

**Comment by carlocab** [comment]
>The test workflow should only rebuild a Linux bottle in the case there was previously a Linux bottle for that formula. Else it should just ignore that part of the build. Can we achieve this with test-bot?

`test-bot` already does this for ARM; I imagine the same logic could be reused for Linux.

File: .github/workflows/linux-pr-tests.yml
Comment (iMichka) [reply .github/workflows/linux-pr-tests.yml:111]: Done. I aligned the code with the one used in the Mac workflow. Thanks.

**Comment by iMichka** [comment]
Here we go: https://github.com/Homebrew/homebrew-test-bot/pull/579

It's inspired from the ARM code.

**Comment by dawidd6** [comment]
Why `linux-pr-tests` and not `linux-tests`? It breaks the consistency with dispatch workflow a bit.

File: .github/workflows/linux-tests.yml
Comment (carlocab) [inline .github/workflows/linux-tests.yml:21]:
```suggestion
      # Will be built in the linux-pr-tests-wheezy workflow
```
Should those also be called `linux-tests-wheezy`?

**Comment by iMichka** [comment]
I renamed the workflows

**Comment by iMichka** [comment]
BTW I did not re-enable the bottle uploading part for the wheezy tests, I'll do that in a second step. And I think that workflow will fail to build a bottle anyway because `binutils` needs `zlib`, so we can't build these bottles in homebrew-core right now.

**Comment by iMichka** [comment]
Also, I forgot to ask: will the automerge workflow work after this change? I guess it needs to pull the Linux bottles from this new Linux workflow too.

**Comment by dawidd6** [comment]
Automerge just runs pr-publish which then triggers workflow runs that run pr-pull. Just modify `brew pr-pull --workflows` in publishing workflow file and it should be good.

**Comment by dawidd6** [comment]
Don't we need a similar step for Linux?

```yml
      - name: Check whether ARM tests will be forced
        if: needs.tap_syntax.outputs.skip-arm-tests
        run: echo 'HOMEBREW_REQUIRE_BOTTLED_ARM=1' >> $GITHUB_ENV
```

Just the env var would be `HOMEBREW_REQUIRE_BOTTLED_LINUX`.

**Comment by dawidd6** [comment]
Also we might need to temporarily add:

```
HOMEBREW_BOTTLE_DOMAIN: https://homebrew.bintray.com/
```

Like it is in dispatch workflow. Without this, the bottles are pulled from linuxbrew.

File: .github/workflows/linux-tests.yml
Comment (dawidd6) [inline .github/workflows/linux-tests.yml:61]: I believe this step is not needed on Linux.

File: .github/workflows/linux-tests.yml
Comment (dawidd6) [inline .github/workflows/linux-tests.yml:120]: This is unneeded too. GitHub runners are ephemeral.

File: .github/workflows/linux-tests.yml
Comment (dawidd6) [inline .github/workflows/linux-tests.yml:123]: Same as above.

File: .github/workflows/linux-tests.yml
Comment (dawidd6) [inline .github/workflows/linux-tests.yml:54]: Since we don't run the doctor step, this env variable and HOMEBREW_FORCE_HOMEBREW_ON_LINUX might not be needed because setup-homebrew already configures the right tap for us. Speaking this after creating the dispatch workflow, though I'm not fully sure.

File: .github/workflows/linux-tests.yml
Comment (dawidd6) [inline .github/workflows/linux-tests.yml:55]: Ye, I think we should drop this.

https://github.com/Homebrew/brew/blob/a65c5d685c1988007882bdbb562799df6e1d4f3d/Library/Homebrew/global.rb#L91-L92

File: .github/workflows/linux-tests-wheezy.yml
Comment (MikeMcQuaid) [inline .github/workflows/linux-tests-wheezy.yml:1]:
```suggestion
name: GitHub Actions CI (Linux Wheezy)
```

File: .github/workflows/linux-tests.yml
Comment (MikeMcQuaid) [inline .github/workflows/linux-tests.yml:1]:
```suggestion
name: GitHub Actions CI (Linux)
```

File: .github/workflows/linux-tests.yml
Comment (MikeMcQuaid) [inline .github/workflows/linux-tests.yml:4]: Is there any way of extracting this to a list that can be shared somewhere?

File: .github/workflows/linux-tests.yml
Comment (MikeMcQuaid) [reply .github/workflows/linux-tests.yml:1]: I think it'd be good for us to figure out a way to unify this with `tests.yml`.

**Review by MikeMcQuaid — COMMENTED** [review]
After this PR is merged does that mean we'll start running Linux CI on every PR? I don't think we're ready for that yet until we're (very) close to having the two repositories entirely in sync.

I think a happy middle ground would be adding some more specific formulae which are entirely in sync and updated regularly to be run on Linux as a test so we're not introducing more builds/failures that will confuse contributors.

File: .github/workflows/linux-tests.yml
Comment (dawidd6) [reply .github/workflows/linux-tests.yml:4]: With `paths-ignore` no. With another job running before tests and determining if they should be run, maybe, but then we would get the "skipped" spam.

File: .github/workflows/linux-tests-wheezy.yml
Comment (dawidd6) [reply .github/workflows/linux-tests-wheezy.yml:1]: Looks like my force push containing these changes was overriden :(.

File: .github/workflows/linux-tests.yml
Comment (iMichka) [reply .github/workflows/linux-tests.yml]: Done (changed the name)

**Comment by iMichka** [comment]
> After this PR is merged does that mean we'll start running Linux CI on every PR?

Yes, but test-bot should not do much as I merged https://github.com/Homebrew/homebrew-test-bot/pull/579, so if there is no bottle nothing will be built.

This will allow to add bottles step by step.

I do not plan to add bottles before we have the self-hosted runner, in case we have a long running job.

**Review by carlocab — APPROVED** [review]

File: .github/workflows/linux-tests.yml
Comment (MikeMcQuaid) [reply .github/workflows/linux-tests.yml:4]: @dawidd6 Ok. I think this is fine for now but eventually I think the "skipped" would be preferable.

**Comment by MikeMcQuaid** [comment]
> Yes, but test-bot should not do much as I merged [Homebrew/homebrew-test-bot#579](https://github.com/Homebrew/homebrew-test-bot/pull/579), so if there is no bottle nothing will be built.

I'd like us to avoid running Linux CI on every PR until we're ready to complete the merge. Doing so on a subset of formulae for now feels like a good fit.

> This will allow to add bottles step by step.

Aren't we skipping the build unless there's an existing bottle?

> I do not plan to add bottles before we have the self-hosted runner, in case we have a long running job.

"all bottles" or "any bottles"? If the latter: that's an even stronger argument for holding off running on every PR.

**Comment by MikeMcQuaid** [comment]
> Why? It's running an "empty" build right now that does nothing. As @dawidd6 said, it's a good first step to test that at least the workflow is setup.

We can still do that test with `workflow_dispatch` and on a subset of formulae/PRs without needing to have it on all of them.

It's going to be noisy and, when it fails, it's going to set (incorrectly) expectations for contributors to fix these problems.

> Yes. That's what we want.

So: this is going to be skipping every single build until we add bottles but still display a CI status? That seems like a bad call. Again: let's maintain a list of specific formulae to run on PRs until we actually have bottles in here that we're needing to maintain.

> Any bottles. If we have a long running build without self hosted runner, we will be stuck, and I don't want this to happen.

So what's the point in this workflow being added at all?

> I have trouble to understand why the bottling strategy that I want to use here was "allowed" for ARM but you are reluctant to use it for Linux. We are basically doing exactly the same, and it worked out well.

A few (big) reasons:
- Non-trivial numbers of people were already using homebrew-core on ARM before we had a single bottle. No-one is using homebrew-core on Linux, they are still using linuxbrew-core. We don't seem to be particularly close to switching over here, either.
- When we enabled ARM CI we were already at >1000 bottles: https://github.com/Homebrew/homebrew-core/pull/67577

---

What I think we should do:

1. add Linux CI using dispatch workflow and test that now
2. wire up this CI to actually upload and publish bottles which are added to homebrew-core formulae bottle blocks
3. add an allowlist for specific, frequently updated, bottled (or `bottle :unneeded`) formulae which have been fully merged from linuxbrew-core
4. complete linuxbrew-core merge
5. enable CI for all formulae in homebrew-core
6. add an environment variable so maintainers can start using homebrew-core instead of linuxbrew-core on Linux
7. bottle everything
8. switch users over to using homebrew-core instead of linuxbrew-core
9. archive linuxbrew-core

**Pull request closed without merging**
